apis/common/parent/projectlocation.go

```diff
@@ -0,0 +1,88 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package parent
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var _ Parent = &ProjectAndLocationParent{}
+
+type ProjectAndLocationParent struct {
+	ProjectID string
+	Location  string
+}
+
+func (p *ProjectAndLocationParent) String() string {
+	return "projects/" + p.ProjectID + "/locations/" + p.Location
+}
+
+func (p *ProjectAndLocationParent) MatchActual(actualI Parent) error {
+	actual, ok := actualI.(*ProjectAndLocationParent)
+	if !ok {
+		return fmt.Errorf("parent format changed, desired %T", p)
+	}
+	if p.ProjectID != actual.ProjectID {
+		return fmt.Errorf("spec.projectRef changed, desired %s, actual %s", p.ProjectID, actual.ProjectID)
+	}
+	if p.Location != actual.Location {
+		return fmt.Errorf("spec.location changed, desired %s, actual %s", p.Location, actual.Location)
+	}
+	return nil
+}
+
+var _ ParentBuilder = &ProjectAndLocationRef{}
+
+// ProjectAndLocationParent specifies the resource's GCP hierarchy (Project/Folder/Organization) and its geographical location.
+// +kubebuilder:object:generate:=true
+type ProjectAndLocationRef struct {
```

Comment:
I think this is actually ProjectAndLocationIdentity, because we want to embed this into FooIdentity, which gets embedded into Foo